feat(chat-models): Tool calling feature for Ollama (#288)

Ganeshsivakumar opened the pull request:
Resolves #288

davidmigloz reviewed (commented):
Overall, it was great 🙌 My comments are mainly improvements. I'll push a new commit resolving them.

davidmigloz commented on packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart:
````diff
@@ -0,0 +1,210 @@
+///},
+///);
+///```
+class OllamaTools extends BaseChatModel<ChatOllamaToolOptions> {
````
fixed

davidmigloz commented on packages/langchain_ollama/lib/src/chat_models/types.dart:
```diff
@@ -271,3 +274,62 @@ class ChatOllamaOptions extends ChatModelOptions {
+/// {@template chat_ollama_tool_options}
+/// Options to pass into OllamaTools.
+/// {@endtemplate}
+class ChatOllamaToolOptions extends ChatModelOptions {
```
I've split it into:
- `chat_ollama`
- `chat_ollama_tools`

davidmigloz commented on packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart:
```diff
@@ -0,0 +1,210 @@
+          defaultOptions.options.model ??
+          throwNullModelError(),
+      messages: messages.toMessages(),
+      format: options?.format?.toResponseFormat(),
```
done

davidmigloz commented on packages/langchain_ollama/test/chat_models/ollama_tools_test.dart:
```diff
@@ -0,0 +1,126 @@
+        },
+      );
+      test(
+          'should return default tool if no other tools should be called for a given query',
```
Now the `_conversational_response` tool is never returned. If the model calls it, the input of the tool is returned as the content of the `AIChatMessage`

davidmigloz commented on packages/langchain_ollama/lib/src/chat_models/types.dart:
```diff
@@ -271,3 +274,62 @@ class ChatOllamaOptions extends ChatModelOptions {
+
+  ///Default system template that will be used
+  static const String defaultToolSystemPromtTemplate =
+      'You have access to the following tools: {tools} You must always select one of the above tools based on question and respond with only a JSON object matching the following schema:{{"tool": <name of the selected tool>,"tool_input": <parameters for the selected tool, matching the tools JSON schema>}}. Respond using JSON';
```
Based on the `toolChoice` passed, it now customizes the tools specified in the prompt

davidmigloz commented:
Could you take a look at my changes?

Pending things:
- Port [the documentation](https://js.langchain.com/v0.2/docs/integrations/chat/ollama_functions)
- Streaming support (we can leave it outside the scope of this PR)

Ganeshsivakumar commented:
Refactored changes for formatting prompt logic with tool choice option , parsing result, handling multiple tool result and updated folder structure looks good.

davidmigloz commented:
I've added the documentation. It's good to go 🚀
